File: GO/Structures/ListDynamic.go
package Structures

type NodeList struct {
    value int
    next  *NodeList
}

type ListDynamic struct {
    head *NodeList
    tail *NodeList  // ← ADICIONADO: ponteiro para o último elemento
    size int
}

func NewListDynamic() *ListDynamic {
    return &ListDynamic{
        head: nil,
        tail: nil,
        size: 0,
    }
}

func (l *ListDynamic) Add(value int) {
    newNode := &NodeList{value: value, next: nil}
    
    if l.head == nil {
        l.head = newNode
        l.tail = newNode  // ← Primeiro elemento é head e tail
    } else {
        l.tail.next = newNode  // ← O(1) - adiciona direto no final!
        l.tail = newNode       // ← Atualiza tail
    }
    l.size++
}

func (l *ListDynamic) ToArray() []int {
    result := make([]int, l.size)
    current := l.head
    i := 0
    
    for current != nil {
        result[i] = current.value
        current = current.next
        i++
    }
    
    return result
}

func (l *ListDynamic) FromArray(arr []int) {
    l.head = nil
    l.tail = nil  // ← Resetar tail também
    l.size = 0
    
    for _, v := range arr {
        l.Add(v)
    }
}

func (l *ListDynamic) Size() int {
    return l.size
}

File: PYTHON/structures/list_dynamic.py
# structures/list_dynamic.py
from typing import List, Optional
import sys
sys.path.append('..')
from utils.data_structure import DataStructure


class NodeList:
    """Nó para lista encadeada"""
    def __init__(self, value: int):
        self.value = value
        self.next: Optional['NodeList'] = None


class ListDynamic(DataStructure):
    """Lista com alocação dinâmica (lista encadeada)"""
    
    def __init__(self):
        self.head: Optional[NodeList] = None
        self.tail: Optional[NodeList] = None  # Otimização O(1) para inserção
        self._size = 0
    
    def add(self, value: int) -> None:
        new_node = NodeList(value)
        
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        
        self._size += 1
    
    def to_array(self) -> List[int]:
        result = []
        current = self.head
        
        while current is not None:
            result.append(current.value)
            current = current.next
        
        return result
    
    def from_array(self, arr: List[int]) -> None:
        self.head = None
        self.tail = None
        self._size = 0
        
        for v in arr:
            self.add(v)
    
    def size(self) -> int:
        return self._size